src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs

```diff
@@ -0,0 +1,262 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class DataFlowPass
+    {
+        /// <summary>
+        /// Holds info from the assignment pass done by
+        /// <see cref="LocalFunctionAssignmentPass"/>.
+        /// </summary>
+        private LocalFuncAssignmentResults _localFuncAssignmentResults;
+
+        /// <summary>
+        /// True if we're currently recording reads, rather than reporting
+        /// diagnostics.
+        /// </summary>
+        private bool _recordingReads = false;
+
+        private SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> _readVars =
+            new SmallDictionary<LocalFunctionSymbol, LocalFuncInfo>();
+
+
+        protected class LocalFuncInfo
+        {
+            public bool IsDirty { get; set; }
+            public PooledHashSet<int> UsedVars =
+                PooledHashSet<int>.GetInstance();
+
+            public void Free()
+            {
+                UsedVars.Free();
+            }
+        }
+
+        protected class LocalFuncAssignmentResults
+        {
+            public SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> AssignedVars { get; }
+            public ImmutableArray<VariableIdentifier> VariableSlots { get; }
+
+            public LocalFuncAssignmentResults(
+                SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> assignedVars,
+                ImmutableArray<VariableIdentifier> variableSlots)
+            {
+                AssignedVars = assignedVars;
+                VariableSlots = variableSlots;
+            }
+
+            public void Free()
+            {
+                FreeLocalFuncInfos(AssignedVars.Values);
+            }
+        }
+
+        private static void FreeLocalFuncInfos<T>(T infos)
+            where T : IEnumerable<LocalFuncInfo>
+        {
+            foreach (var info in infos)
+            {
+                info.Free();
+            }
+        }
```

Comment:
There's some amount of complexity in freeing the `LocalFuncInfo.UsedVars` collections. Do we need to use pooled `HashSet<T>` here?